libcache/cache.c

```diff
@@ -0,0 +1,391 @@
+	memset(set->lines, 0, sizeof(set->lines));
+
+	if (set->tags == NULL) {
+		free(set);
+		return NULL;
+	}
+
+	return set;
+}
+
+
+/* Generates mask of type uint64_t with numBits set to 1 */
+uint64_t cache_genMask(int numBits)
+{
+	return ((uint64_t)1 << (uint64_t)numBits) - (uint64_t)1;
+}
+
+
+cachectx_t *cache_init(size_t size, size_t lineSize, cache_writeCb_t writeCb, cache_readCb_t readCb)
+{
+	int i = 0, j = 0;
+	cacheset_t *set = NULL;
+
+	cachectx_t *cache = malloc(sizeof(cachectx_t));
+
+	if (cache == NULL) {
+		return NULL;
+	}
+
+	if (size % lineSize != 0 || size % LIBCACHE_NUM_WAYS != 0) {
+		fprintf(stderr, "Invalid size parameters: size in bytes must be multiple of lineSize and multiple of %d\n", LIBCACHE_NUM_WAYS);
+		free(cache);
+		return NULL;
+	}
+
+	cache->numLines = size / lineSize;
+
+	cache->numSets = cache->numLines / LIBCACHE_NUM_WAYS;
+
+	cache->sets = calloc(cache->numSets, sizeof(cacheset_t));
+
+	if (cache->sets == NULL) {
+		free(cache);
+		return NULL;
+	}
+
+	for (; i < cache->numSets; ++i) {
+		set = cache_setInit();
+
+		if (set == NULL) {
+			for (; j < i; ++j) {
+				free(cache->sets[j].tags);
+			}
+			free(cache->sets);
+			free(cache);
+
+			return NULL;
+		}
+		cache->sets[i] = *set;
+
+		cache->sets[i].count = 0;
+
+		free(set);
+	}
+
+	cache->offBitNum = log2(lineSize);
+	off_t setBits = log2(cache->numSets);
+	off_t tagBits = LIBCACHE_ADDR_WIDTH - setBits - cache->offBitNum;
+
+	cache->tagMask = cache_genMask(tagBits);
+	cache->setMask = cache_genMask(setBits);
+	cache->offMask = cache_genMask(cache->offBitNum);
+
+	cache->readCb = readCb;
+	cache->writeCb = writeCb;
+
+	return cache;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i = 0;
+
+	for (; i < cache->numSets; ++i) {
+		if (cache->sets != NULL) {
+			free(cache->sets[i].tags);
+		}
+	}
+
+	free(cache->sets);
+
+	free(cache);
+
+	return 0;
+}
+
+
+int cache_compareTags(const void *lhs, const void *rhs)
+{
+	int ret = 0;
+
+	cacheline_t **lhsLine = (cacheline_t **)lhs;
+	cacheline_t **rhsLine = (cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+ssize_t cache_writeToSet(cache_writeCb_t writeCb, cacheset_t *set, cacheline_t *line, const off_t offset, int policy)
+{
+	int i = 0;
+	ssize_t wrt = 0;
+	cacheline_t *temp = NULL, *oldest = NULL, **linePtr = NULL;
+
+	if (set->count < LIBCACHE_NUM_WAYS) {
+		for (; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (set->lines[i].isValid != '1') {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		temp = &(set->lines[i]);
+
+		switch (policy) {
+			case 0: {
+				/* write-back */
+				set->lines[i].isDirty = '1';
+				break;
+			}
```

Comment:
No need for brackets (needed only if we need to separate code block to e.g. declare new variable, not applicable here)
```suggestion
			case 0:
				/* write-back */
				set->lines[i].isDirty = '1';
				break;
```

Reply:
Thank you, removed